components/builder-api/src/server/resources/origins.rs

```diff
@@ -269,15 +272,108 @@ fn delete_origin(req: HttpRequest, path: Path<String>, state: Data<AppState>) ->
         Err(err) => return err.into(),
     };
 
-    match Origin::delete(&origin, &*conn).map_err(Error::DieselError) {
-        Ok(_) => HttpResponse::NoContent().into(),
+    // Prior to passing the deletion request to the backend, we validate
+    // that the user has already cleaned up the most critical origin data.
+    match origin_delete_preflight(&origin, &*conn) {
+        Ok(_) => {
+            match Origin::delete(&origin, &*conn).map_err(Error::DieselError) {
+                Ok(_) => HttpResponse::NoContent().into(),
+                Err(err) => {
+                    debug!("Origin {} deletion failed! err = {}", origin, err);
+                    // We do not want to expose any database details from diesel
+                    // thus we simply return a 409 with an empty body.
+                    HttpResponse::new(StatusCode::CONFLICT)
+                }
+            }
+        }
         Err(err) => {
-            debug!("Origin {} is not deletable, err = {}", origin, err);
-            HttpResponse::new(StatusCode::CONFLICT)
+            debug!("Origin preflight determined that {} is not deletable, err = {}!",
+                   origin, err);
+            // Here we want to enrich the http response with a sanitized error
+            // by returning a 409 with a helpful message in the body.
+            HttpResponse::with_body(StatusCode::CONFLICT, Body::from_message(format!("{}", err)))
         }
     }
 }
 
+#[allow(clippy::needless_pass_by_value)]
```

Comment:
Is this clippy allow needed?

Reply:
Turns out it's not! I removed in https://github.com/habitat-sh/builder/pull/1298/commits/6d356c4c2c2f130bb43a14b9c87a2c8a19314a8d